cli/commands/terraform/creds/providers/externalcmd/provider.go

```diff
@@ -0,0 +1,110 @@
+package externalcmd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt/cli/commands/terraform/creds/providers"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/shell"
+	"golang.org/x/exp/maps"
+)
+
+// Provider runs external command that returns a json string with credentials.
+type Provider struct {
+	terragruntOptions *options.TerragruntOptions
+}
+
+// NewProvider returns a new Provider instance.
+func NewProvider(opts *options.TerragruntOptions) providers.Provider {
+	return &Provider{
+		terragruntOptions: opts,
+	}
+}
+
+// Name implements providers.Name
+func (provider *Provider) Name() string {
+	return fmt.Sprintf("external %s command", provider.terragruntOptions.AuthProviderCmd)
+}
+
+// GetCredentials implements providers.GetCredentials
+func (provider *Provider) GetCredentials(ctx context.Context) (*providers.Credentials, error) {
+	command := provider.terragruntOptions.AuthProviderCmd
+	if command == "" {
+		return nil, nil
+	}
+
+	var args []string
+	if parts := strings.Fields(command); len(parts) > 1 {
+		command = parts[0]
+		args = parts[1:]
+	}
+
+	output, err := shell.RunShellCommandWithOutput(ctx, provider.terragruntOptions, "", true, false, command, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	if output.Stdout == "" {
+		return nil, errors.Errorf("the command %s completed successfully, but the response does not contain json string", command)
+	}
+
+	type response struct {
+		AWSCredentials *AWSCredentials   `json:"awsCredentials"`
+		Envs           map[string]string `json:"envs"`
+	}
+	resp := &response{Envs: make(map[string]string)}
+
+	if err := json.Unmarshal([]byte(output.Stdout), &resp); err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	creds := &providers.Credentials{
+		Name: providers.AWSCredentials,
+		Envs: resp.Envs,
+	}
+
+	if resp.AWSCredentials != nil {
+		if envs := resp.AWSCredentials.Envs(provider.terragruntOptions); envs != nil {
+			maps.Copy(creds.Envs, envs)
```

Comment:
We should be very explicit in documentation that the same key being present in `providers.AWSCredentials` and `resp.Envs` will result in the `providers.AWSCredentials` taking precedence.

Reply:
Agree. Thanks for the doc suggestion!